content/general/authentication.textile

```diff
@@ -139,22 +141,91 @@ ably.requestToken({ clientId: 'client@example.com' }, function(err, token) {
 
-"Tokens":#tokens are issued from an existing "API key":https://support.ably.io/solution/articles/3000030054-what-is-an-app-api-key, and their capabilities can, at most, match the capabilities of the issuing API key. If an API key must be shared with a third party, then it is recommended that "the principle of least privilege":http://en.wikipedia.org/wiki/Principle_of_least_privilege is considered assigning only the capabilities needed by that third party. Thus, any Ably requests authenticated using that API key or tokens issued from that API key, will be restricted to the capabilities assigned to the key.
+<a href="/images/diagrams/ably-jwt-auth.png" target="_blank">
+  <img src="/images/diagrams/ably-jwt-auth.png" style="width: 100%" alt="Ably JWT auth method">
+</a>
 
-Capabilities for tokens are determined as follows:
+An example of creating an Ably JWT can be seen below:
+
+bc[javascript]. var header = { 
+    "typ":"JWT",
+    "alg":"HS256",
+    "kid": "{{API_KEY_NAME}}"
+  }
+  var claims = {
+    "iat": currentTime,
+    "exp": currentTime + 3600,
+    "x-ably-capability": "{\"*\":[\"*\"]}"
+  }
+  var base64Header = btoa(header);
+  var base64Claims = btoa(claims);
+  /* Apply the hash specified in the header */
+  var signature = hash((base64Header + "." + base64Claims), SECRET);
+  var ablyJwt = base64Header + "." + base64Claims + "." + signature;
+
+*Note:* At present Ably does not support asymmetric signatures based on a keypair belonging to a third party. If this is something you'd be interested in, please "get in touch":https://www.ably.io/contact.
+
+h4(#jwt-embed-process). Ably-compatible token is embedded in a External JWT from your server and passed to clients
+
+If a system has an existing "JWT":https://jwt.io/ scheme, it can be useful to embed an Ably-compatible token as a claim within it. To do this, the External JWT and need to ensure it meets the following requirements:
+
+* The Ably-compatible token is an "Ably JWT":#ably-jwt-process, or an "Ably Token":#token-process
+* The Ably-compatible token is included under the @x-ably-token@ key in the "JOSE Header":https://tools.ietf.org/html/rfc7519
+* OR (if using JWS) the Ably-compatible token is included using the @x-ably-token@ Claim in the payload
+* The expiry time of the Ably-compatible token must not be earlier than the outer JWT’s expiry time (@exp@ Claim). Ably will reject any JWT if it is unencrypted and its @exp@ Claim is later than the expiry of the enclosed Ably-compatible token. This helps to ensure that tokens are renewed prior to expiry
+
+<a href="/images/diagrams/jwt-embed-auth.png" target="_blank">
+  <img src="/images/diagrams/jwt-embed-auth.png" style="width: 100%" alt="Token embedded in a JWT auth method">
```

Comment:
If we had to choose one of the two ways presented to users, I am not sure we should really encourage people to get Ably tokens on their servers and embed them in the JWT.  Would they not be better embedding an Ably JWT token in their token as a) they have the JWT library already, b) it saves an unnecessary request to our servers from theirs.  @paddybyers @tomczoink WDYT?

Reply:
I realise too many options is a bad thing, but could we consider explaining the pros & cons of each at least?